### PR TITLE
Test refactoring leah

### PR DESCRIPTION
**Pull request by nonsensicle**
## CommentServlet & UserServlet
- Remove direct access to Firestore methods and makes other slight changes for testability
## Auth
- Add getUid() method (encompasses decoded FirebaseToken to String logic)
## CommentServletTest & UserServletTest
- Add comprehensive tests for both refactored servlets
## DBInterface.java & FirebaseDB.java
- Add isCreatedComment() method

### CODE REVIEW

**Comment by nonsensicle** [comment]
Haven't pushed the comment tests yet but this is enough to be getting on with for review

File: src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 1]
```diff
@@ -58,52 +74,55 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
     }
-    newComment.creatorId = decodedToken.getUid();
-    ApiFuture addCommentFuture = DBUtils.comments(recipeID).document().set(newComment);
-    DBUtils.blockOnFuture(addCommentFuture);
+    String uid = Auth.getUid(token, response);
+    if (uid == null) {
+      return;
+    }
+
+    newComment.creatorId = uid;
```
Comment (bradleypickard) [inline src/main/java/com/google/sps/servlets/CommentServlet.java:95]: Here might be a good place to initialize other Comment properties, like timestamp, creator ldap, and votes

File: src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 2]
```diff
@@ -58,52 +74,55 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+    }
+
+    String token = request.getParameter("token");
+    String uid = Auth.getUid(token, response);
+    if (uid == null) {
+      return;
+    }
+
+    if (!db.isCreatedComment(recipeID, commentID, uid)) {
+      response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
```
Comment (bradleypickard) [inline src/main/java/com/google/sps/servlets/CommentServlet.java:119]: In other places we set the status as SC_FORBIDDEN if the user is not the creator. Doesn't really matter much but it's helpful in debugging

File: src/main/java/com/google/sps/servlets/UserServlet.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    String token = request.getParameter("token");
+    boolean isSaveRequest = Boolean.parseBoolean(request.getParameter("saved"));
+
+    String uid = Auth.getUid(token, response);
+    if (uid == null) {
+      // Auth sets response status to unauthorized.
+      return;
+    }
+
+    if (isSaveRequest) {
```
Comment (hanpeiz) [inline src/main/java/com/google/sps/servlets/UserServlet.java:86]: I think you should have separate helper function for each type of "subrequest". You can have an enum for each request type in your HttpServletRequest, and process each separate request in a switch statement.

something like:
```
RequestType request = RequestType.valueOf(request.getParameter("type"));
Switch (request):
  case RequestType.SAVE:
    saveRecipe(recipeID, UserId);
    break;
  case RequestType.UPDATE_USERNAME:
    .....
```

File: src/main/webapp/feed.html [hunk 4]
```diff
@@ -33,18 +42,116 @@ <h1>
...
+      // Performs retrieval of recipes based on the filters currently contained in <div> 'selected-filters'
+      function getRecipes() {
+        const tagIDs = getCurrentFilters()
+        let recipeList = document.getElementById("recipe-list")
+        let queryString = "/api/post"
+
+        if (tagIDs.length > 0) {
+            queryString += "?tagIDs=" + tagIDs[0]
+            for (var x = 1; x < tagIDs.length; x++) {
+                queryString += "&tagIDs=" + tagIDs[x];
```
Comment (hanpeiz) [inline src/main/webapp/feed.html:85]: I think you can use join() here with &tagIDs= as the separator.

File: src/main/webapp/profile.html [hunk 5]
```diff
@@ -17,5 +33,125 @@
...
+        } else {
+          window.location.href = "/login-page.html"
+        }
+      }
+
+      // TODO:
+      /** Retrieves a user's saved posts. */
+      function getUserSavedRecipes(idToken) {
+        fetch("/api/post?token=" + idToken + "&saved=true").then(async (response) => {
+          if (response.status < 200 || response.status >= 300) {
```
Comment (hanpeiz) [inline src/main/webapp/profile.html:95]: !response.ok

File: src/main/webapp/feed.html (hunk 4 above)
Comment (nonsensicle) [reply src/main/webapp/feed.html:85]: @bradleypickard  i might need a little help on this one

Comment (hari387) [reply src/main/webapp/feed.html:85]:
```
qsTags = tagIDs.map(tag => "tagIds=" + tag)
queryString += "?" + qsTags.join("&")
```
think that should do it

File: src/main/java/com/google/sps/servlets/CommentServlet.java (hunk 2 above)
Comment (nonsensicle) [reply src/main/java/com/google/sps/servlets/CommentServlet.java]: thanks!

**Comment by hari387** [comment]
closes #6 , closes #43 , closes #44 , closes #29

**Pull request merged**